File: mcp-code-intelligence-kotlin/src/test/kotlin/com/codeintel/memory/search/TokenBudgetTest.kt
/** Unit tests for TokenBudget — search result token limiting. */
package com.codeintel.memory.search

import com.codeintel.memory.models.KnowledgeEntry
import com.codeintel.memory.models.KnowledgeSearchResult
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertFalse
import kotlin.test.assertTrue

class TokenBudgetTest {

    private fun makeResult(content: String, id: Long = 1): KnowledgeSearchResult {
        return KnowledgeSearchResult(
            entry = KnowledgeEntry(id = id, content = content, summary = "s", type = "CONTEXT"),
            score = 1.0
        )
    }

    @Test
    fun `returns all results when within budget`() {
        val budget = TokenBudget()
        val results = listOf(
            makeResult("short content", 1),
            makeResult("another short", 2),
        )
        val br = budget.apply(results, 2000)
        assertEquals(2, br.results.size)
        assertFalse(br.truncated)
        assertEquals(2, br.totalMatches)
    }

    @Test
    fun `truncates when budget exceeded`() {
        val budget = TokenBudget()
        val longContent = "x".repeat(4000) // ~1000 tokens
        val results = listOf(
            makeResult(longContent, 1),
            makeResult(longContent, 2),
            makeResult(longContent, 3),
        )
        val br = budget.apply(results, 1500)
        assertTrue(br.results.size < 3)
        assertTrue(br.truncated)
        assertEquals(3, br.totalMatches)
        assertTrue(br.tokensUsed <= 1500)
    }

    @Test
    fun `truncates individual entry to fit remaining budget`() {
        val budget = TokenBudget()
        val results = listOf(
            makeResult("a".repeat(400), 1),  // 100 tokens
            makeResult("b".repeat(4000), 2), // 1000 tokens — won't fit in remaining
        )
        val br = budget.apply(results, 300)
        assertEquals(2, br.results.size)
        assertTrue(br.truncated)
        // Second result should be truncated
        assertTrue(br.results[1].entry.content.length < 4000)
        assertTrue(br.results[1].entry.content.endsWith("..."))
    }

    @Test
    fun `skips entry if remaining budget too small`() {
        val budget = TokenBudget()
        val results = listOf(
            makeResult("a".repeat(1000), 1), // 250 tokens
            makeResult("b".repeat(4000), 2), // 1000 tokens
        )
        // Budget = 260 → first fits (250), remaining = 10 < 50 → skip second
        val br = budget.apply(results, 260)
        assertEquals(1, br.results.size)
        assertTrue(br.truncated)
    }

    @Test
    fun `empty results returns empty`() {
        val budget = TokenBudget()
        val br = budget.apply(emptyList(), 2000)
        assertEquals(0, br.results.size)
        assertFalse(br.truncated)
        assertEquals(0, br.tokensUsed)
    }

    @Test
    fun `countTokens approximation works`() {
        assertEquals(1, countTokens("abcd"))    // 4 chars = 1 token
        assertEquals(2, countTokens("abcde"))   // 5 chars ≈ 2 tokens
        assertEquals(25, countTokens("x".repeat(100))) // 100 chars = 25 tokens
    }

    @Test
    fun `truncateToFit respects token limit`() {
        val text = "x".repeat(1000)
        val truncated = truncateToFit(text, 50) // 50 tokens = 200 chars
        assertTrue(truncated.length <= 200)
        assertTrue(truncated.endsWith("..."))
    }

    @Test
    fun `truncateToFit returns original if within limit`() {
        val text = "short text"
        val result = truncateToFit(text, 100)
        assertEquals(text, result)
    }
}

File: mcp-code-intelligence-nodejs/src/memory/tests/token-budget.test.ts
/**
 * Unit tests for TokenBudget — search result token limiting.
 */

import { describe, it } from 'node:test';
import assert from 'node:assert/strict';
import { TokenBudget } from '../v2/token-budget.js';
import { SearchResult, KnowledgeEntry } from '../models.js';

function makeResult(id: number, contentLength: number): SearchResult {
  const entry: KnowledgeEntry = {
    id, content: 'A'.repeat(contentLength), summary: 'test', type: 'CONTEXT',
    tier: 'WORKING', source: null, source_ref: null, tags: '', confidence: 1,
    access_count: 0, created_at: '', updated_at: '', last_accessed_at: null, expires_at: null,
  };
  return { entry, score: 1.0 - id * 0.1, matchType: 'hybrid' };
}

describe('TokenBudget', () => {
  const budget = new TokenBudget();

  describe('apply', () => {
    it('returns all results when within budget', () => {
      const results = [makeResult(1, 100), makeResult(2, 100)]; // 50 tokens total
      const { results: limited, tokensUsed, truncated } = budget.apply(results, 2000);
      assert.equal(limited.length, 2);
      assert.equal(tokensUsed, 50); // 200 chars / 4
      assert.equal(truncated, false);
    });

    it('limits results when exceeding budget', () => {
      const results = [
        makeResult(1, 4000), // 1000 tokens
        makeResult(2, 4000), // 1000 tokens
        makeResult(3, 4000), // 1000 tokens
      ];
      const { results: limited, truncated, totalMatches } = budget.apply(results, 2000);
      assert(limited.length < 3);
      assert.equal(truncated, true);
      assert.equal(totalMatches, 3);
    });

    it('truncates last entry to fit remaining budget', () => {
      const results = [
        makeResult(1, 4000), // 1000 tokens
        makeResult(2, 8000), // 2000 tokens — won't fully fit
      ];
      const { results: limited, tokensUsed } = budget.apply(results, 1500);
      assert.equal(limited.length, 2);
      assert(tokensUsed <= 1510); // small overhead from "..." suffix
      assert(limited[1].entry.content.length < 8000);
    });

    it('skips entry if remaining budget < 50 tokens', () => {
      const results = [
        makeResult(1, 7900), // 1975 tokens
        makeResult(2, 4000), // 1000 tokens — remaining < 50
      ];
      const { results: limited, truncated } = budget.apply(results, 2000);
      assert.equal(limited.length, 1);
      assert.equal(truncated, true);
    });

    it('handles empty results', () => {
      const { results: limited, tokensUsed, truncated } = budget.apply([], 2000);
      assert.equal(limited.length, 0);
      assert.equal(tokensUsed, 0);
      assert.equal(truncated, false);
    });

    it('handles single result exceeding budget', () => {
      const results = [makeResult(1, 20000)]; // 5000 tokens
      const { results: limited, truncated } = budget.apply(results, 500);
      assert.equal(limited.length, 1);
      assert(limited[0].entry.content.length < 20000);
      assert.equal(truncated, true);
    });

    it('preserves result order (highest score first)', () => {
      const results = [makeResult(1, 400), makeResult(2, 400), makeResult(3, 400)];
      const { results: limited } = budget.apply(results, 2000);
      assert.deepEqual(limited.map(r => r.entry.id), [1, 2, 3]);
    });
  });
});
